### PR TITLE
refactor(query): new hyperloglog and ndv

### LINKED ISSUE #14473: Improve ndv accuracy
**Summary**

### PR DESCRIPTION
**Pull request by sundy-li**
I hereby agree to the terms of the CLA available at: https://docs.databend.com/dev/policies/cla/

### New HyperLogLog implementation

[simple_hll](https://crates.io/crates/simple_hll) is a simple HyperLogLog implementation in rust. It is designed to be simple to use and less space to store (with Sparse HyperLogLog).

### Adaptive Sparse serialization of hyperloglog 

```
#[derive(serde::Deserialize, borsh::BorshDeserialize)]
enum HyperLogLogVariant<const P: usize> {
    Empty,
    Sparse { data: Vec<(u16, u8)> },
    Full(Vec<u8>),
}
```

The serialized bytes could be `[1, 1<<14 = 16k]` with `P=14`.

### Refactor table function `fuse_statistic` and test the ndv results within expected error rate range

```
query T
select *, (ndv - 7500) / 7500 < 0.01625 * 6  from fuse_statistic('db_09_0020', 'a')
----
0 7637 1
1 7730 1
2 7530 1
```

### Additional param to set the error rate in approx_count_distinct function

` approx_count_distinct(0.1)(number)` 

### Refactor column stats using hyperloglog with `P = 12` with max size = 4k (1<<12)

```
/// Takes at most `1<<12 = 4k`` spaces with error ratio of `0.01625`
pub type ColumnStatHLL = HyperLogLog<12>;
```

## Acknowledgements

Some codes and tests are borrowed and inspired from:
- [redis](https://github.com/redis/redis/blob/4930d19e70c391750479951022e207e19111eb55/src/hyperloglog.c)
- [datafusion](https://github.com/apache/arrow-datafusion/blob/f203d863f5c8bc9f133f6dd9b2e34e57ac3cdddc/datafusion/physical-expr/src/aggregate/hyperloglog.rs)
- [pdatastructs](https://github.com/crepererum/pdatastructs.rs/blob/3997ed50f6b6871c9e53c4c5e0f48f431405fc63/src/hyperloglog.rs)

Reference papers:
- [New cardinality estimation algorithms for HyperLogLog sketches](https://arxiv.org/abs/1702.01284)

Thanks @Jimexist  @crepererum  for the initial codes and the paper author: Otmar Ertl

## Others
- Fixes #14473

## Tests

- [ ] Unit Test
- [x] Logic Test
- [ ] Benchmark Test
- [ ] No Test  - _Explain why_

## Type of change

- [ ] Bug Fix (non-breaking change which fixes an issue)
- [ ] New Feature (non-breaking change which adds functionality)
- [ ] Breaking Change (fix or feature that could cause existing functionality not to work as expected)
- [ ] Documentation Update
- [x] Refactoring
- [ ] Performance Improvement
- [ ] Other (please describe):

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/14585)

### CODE REVIEW

**Review by lichuang — APPROVED** [review]

**Comment by sundy-li** [comment]
I found the join order is not very accurate (due to inaccurate selectivity) , we need support #14587 later .

File: src/query/storages/parquet/src/parquet2/statistics.rs [hunk 1]
```diff
@@ -104,7 +104,7 @@ impl BatchStatistics {
             unsafe { self.max_values.index_unchecked(index).to_owned() },
             self.null_count[index],
             0, // this field is not used.
-            self.distinct_count.as_ref().map(|d| d[index]),
+            None,
         )
```
Comment (Dousir9) [inline src/query/storages/parquet/src/parquet2/statistics.rs:108]: Do we need to call `with_distinct_of_values` with `self.distinct_count` here ?

Comment (sundy-li) [reply src/query/storages/parquet/src/parquet2/statistics.rs:108]: Yes, but it's a little trivial.

1. pruning doesn't need the `distinct_of_value` field.
2. parquet2 is being deprecated soon.

Comment (Dousir9) [reply src/query/storages/parquet/src/parquet2/statistics.rs:108]: Got it.

**Review by Dousir9 — APPROVED** [review]

File: src/query/service/src/test_kits/check.rs [hunk 2]
```diff
@@ -75,27 +74,23 @@ pub async fn check_data_dir(
     fixture: &TestFixture,
     case_name: &str,
     snapshot_count: u32,
-    table_statistic_count: u32,
+    _table_statistic_count: u32,
```
Comment (xudong963) [inline src/query/service/src/test_kits/check.rs:77]: Can the arg be removed?

File: src/query/service/src/test_kits/check.rs [hunk 3]
```diff
@@ -135,11 +127,11 @@ pub async fn check_data_dir(
...
-        ts_count, table_statistic_count,
-        "case [{}], check snapshot statistics count",
-        case_name
-    );
+    // assert_eq!(
+    //     ts_count, table_statistic_count,
+    //     "case [{}], check snapshot statistics count",
+    //     case_name
+    // );
     assert_eq!(
```
Comment (xudong963) [inline src/query/service/src/test_kits/check.rs:135]: useless?

**Review by xudong963 — APPROVED** [review]

**Comment by sundy-li** [comment]
The segment file may grow too much.  Converted to draft now.

**Pull request closed without merging**